iceoryx_posh/include/iceoryx_posh/internal/popo/base_server.hpp

```diff
@@ -0,0 +1,154 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef IOX_POSH_POPO_BASE_SERVER_HPP
+#define IOX_POSH_POPO_BASE_SERVER_HPP
+
+#include "iceoryx_hoofs/cxx/expected.hpp"
+#include "iceoryx_posh/capro/service_description.hpp"
+#include "iceoryx_posh/internal/popo/ports/server_port_user.hpp"
+#include "iceoryx_posh/popo/server_options.hpp"
+#include "iceoryx_posh/popo/trigger_handle.hpp"
+#include "iceoryx_posh/runtime/posh_runtime.hpp"
+
+namespace iox
+{
+namespace popo
+{
+using uid_t = UniquePortId;
+
+/// @brief The BaseServer class contains the common implementation for the different server
+/// @param[in] Port type of the underlying port, required for testing specializations.
+template <typename Port = ServerPortUser>
+class BaseServer
+{
+  public:
+    virtual ~BaseServer() noexcept;
+
+    BaseServer(const BaseServer& other) = delete;
+    BaseServer& operator=(const BaseServer&) = delete;
+    BaseServer(BaseServer&& rhs) = delete;
+    BaseServer& operator=(BaseServer&& rhs) = delete;
+
+    ///
+    /// @brief Get the UID of the server.
+    /// @return The server's UID.
+    ///
+    uid_t getUid() const noexcept;
+
+    ///
+    /// @brief Get the service description of the server.
+    /// @return The service description.
+    ///
+    capro::ServiceDescription getServiceDescription() const noexcept;
+
+    ///
+    /// @brief Offer the service to be connected to.
+    ///
+    void offer() noexcept;
+
+    ///
+    /// @brief Stop offering the service.
+    ///
+    void stopOffer() noexcept;
+
+    ///
+    /// @brief Check if the server is offering.
+    /// @return True if service is currently being offered.
+    ///
+    bool isOffered() const noexcept;
+
+    ///
+    /// @brief Check if the server has clients
+    /// @return True if currently has subscribers to the service.
+    ///
+    bool hasClients() const noexcept;
+
+    ///
+    /// @brief Check if requests are available.
+    /// @return True if requests are available.
+    ///
+    bool hasRequests() const noexcept;
+
+    ///
+    /// @brief Check if requests has been missed since the last call of this method.
+    /// @return True if requests has been missed.
+    /// @details Requests may be missed due to overflowing receive queue.
+    ///
+    bool hasMissedRequests() noexcept;
+
+    /// @brief Releases any unread queued requests.
+    void releaseQueuedRequests() noexcept;
+
+    friend class NotificationAttorney;
+
+  protected:
+    using SelfType = BaseServer<Port>;
+    using PortType = Port;
+
+    BaseServer(const capro::ServiceDescription& service, const ServerOptions& serverOptions) noexcept;
+
+    /// @brief Only usable by the WaitSet, not for public use. Invalidates the internal triggerHandle.
```

Comment:
Not important but again: WaitSet and Listener for all the event related stuff and the state related stuff is only supported by the waitset